chain/chain/src/doomslug.rs

```diff
@@ -0,0 +1,1004 @@
+    ///     produce a block (in practice a blocks might not be produceable yet if not enough time
+    ///     passed since it accumulated enough approvals)
+    pub fn get_largest_height_crossing_threshold(&self) -> BlockHeight {
+        self.largest_threshold_height
+    }
+
+    pub fn get_largest_height_with_doomslug_finality(&self) -> BlockHeight {
+        self.largest_ds_final_height
+    }
+
+    pub fn get_largest_skipped_height(&self) -> BlockHeight {
+        self.largest_promised_skip_height
+    }
+
+    pub fn get_largest_endorsed_height(&self) -> BlockHeight {
+        self.largest_endorsed_height
+    }
+
+    pub fn get_timer_height(&self) -> BlockHeight {
+        self.timer.height
+    }
+
+    pub fn get_timer_start(&self) -> Instant {
+        self.timer.started
+    }
+
+    /// Is expected to be called periodically and processed the timer (`start_timer` in the paper)
+    /// If the `cur_time` way ahead of last time the `process_timer` was called, will only process
+    /// a bounded number of steps, to avoid an infinite loop in case of some bugs.
+    /// Processes sending delayed approvals or skip messages
+    ///
+    /// # Arguments
+    /// * `cur_time` - is expected to receive `now`. Doesn't directly use `now` to simplify testing
+    ///
+    /// # Returns
+    /// A vector of approvals that need to be sent to other block producers as a result of processing
+    /// the timers
+    #[must_use]
+    pub fn process_timer(&mut self, cur_time: Instant) -> Vec<Approval> {
+        let mut ret = vec![];
+        for _i in 0..MAX_TIMER_ITERS {
+            let delay = self
+                .timer
+                .get_delay(self.timer.height.saturating_sub(self.largest_ds_final_height));
+            if cur_time >= self.timer.started + delay {
+                if self.timer.height > self.largest_endorsed_height
+                    && self.timer.height > self.tip.height
+                {
+                    self.largest_promised_skip_height =
+                        std::cmp::max(self.timer.height, self.largest_promised_skip_height);
+                    if let Some(approval) = self.create_approval(self.timer.height + 1, false) {
+                        ret.push(approval);
+                    }
+                }
+
+                // Restart the timer
+                self.timer.started += delay;
+                self.timer.height += 1;
+            } else {
+                break;
+            }
+        }
+
+        ret
+    }
+
+    pub fn create_approval(
+        &self,
+        target_height: BlockHeight,
+        is_endorsement: bool,
+    ) -> Option<Approval> {
+        self.me.as_ref().map(|me| {
+            Approval::new(
+                self.tip.block_hash,
+                self.tip.reference_hash,
+                target_height,
+                is_endorsement,
+                &**self.signer.as_ref().unwrap(),
```

Comment:
this looks like something evil.
```
if let Some((me, signer)) = (self.me.as_ref(), self.signer.as_ref()) {
   ... 
    &*signer
}
```

also might be easier to pack `me` + `signer` into one object - BlockSigner.
We also need `BlockSigner` to get HSM signing going (because we will need to pass non-hashed data to HSM for signatures anyway ,so we need API that has specific `sign_block`, `sign_approval` APIs).
/\ this all sounds like a separate PR if you are not me :D

Reply:
I don't want to change anything related to HSM yet, it will be a separate effort.

Your approach still requires `&**signer`, because it's a reference to an `Arc`.